packages/common/src/api/tan-query/jupiter/useSwapTokens.ts

```diff
@@ -71,207 +99,74 @@ export const useSwapTokens = () => {
 
   return useMutation<SwapTokensResult, Error, SwapTokensParams>({
     mutationFn: async (params): Promise<SwapTokensResult> => {
-      const {
-        inputMint: inputMintUiAddress,
-        outputMint: outputMintUiAddress,
-        amountUi
-      } = params
-      const slippageBps = params.slippageBps ?? 50
-      const wrapUnwrapSol = params.wrapUnwrapSol ?? true
+      const { inputMint: inputMintUiAddress, outputMint: outputMintUiAddress } =
+        params
 
-      let quoteResult: JupiterQuoteResult | undefined
-      let signature: string | undefined
       let errorStage = 'UNKNOWN'
-      const instructions: TransactionInstruction[] = []
+      let firstQuoteResult: JupiterQuoteResult | undefined
+      let secondQuoteResult: JupiterQuoteResult | undefined
+      let signature: string | undefined
 
       try {
-        // ---------- 1. Initialize Dependencies & Wallet ----------
+        // Initialize dependencies
         errorStage = 'WALLET_INITIALIZATION'
-        const [sdk, keypair] = await Promise.all([
-          audiusSdk(),
-          solanaWalletService.getKeypair()
-        ])
+        const dependenciesResult = await initializeSwapDependencies(
+          solanaWalletService,
+          audiusSdk,
+          queryClient,
+          user
+        )
 
-        if (!keypair) {
-          return {
-            status: SwapStatus.ERROR,
-            error: {
-              type: SwapErrorType.WALLET_ERROR,
-              message: 'Wallet not initialised'
-            }
-          }
+        if ('error' in dependenciesResult) {
+          return dependenciesResult.error
         }
-        const userPublicKey = keypair.publicKey
-        const feePayer = await sdk.services.solanaClient.getFeePayer()
-        const ethAddress = user?.wallet
 
-        // ---------- 2. Prepare Token Information ----------
-        // Find input and output tokens from our backend-driven token data
-        const inputToken = findTokenByAddress(tokens, inputMintUiAddress)
-        const outputToken = findTokenByAddress(tokens, outputMintUiAddress)
+        const dependencies = dependenciesResult
+
+        // Check if this is a direct swap involving AUDIO
+        const isInputAudio = inputMintUiAddress === AUDIO_MINT
+        const isOutputAudio = outputMintUiAddress === AUDIO_MINT
+        const isDirect = isInputAudio || isOutputAudio
```

Comment:
We'll need to have some fancier detection here based on availability. That's why I had written those methods to try with increasing accounts until it worked, etc.

Not necessary from day 1, but something to think about. Hot tokens at some amount of value in the curve will probably have a direct path.